### ISSUE TITLE
is there an equivalent of bash's herestring ?

### ISSUE THREAD

**Issue by adam-devel** [elves/elvish#1261]

**Comment by xiaq** [comment]
Elvish doesn't have it now, although it does look an ergonomic feature in some cases.

**Comment by krader1961** [comment]
It's not obvious adding another feature to the Elvish grammar just to avoid an explicit pipe is worth the cost -- both implementation and documentation. One of the reasons I switched from zsh to fish was I got fed up with additions to the zsh grammar that were seemingly implemented without regard to whether the added complexity was justified and the new feature meshed well with existing features and couldn't be accomplished in some other way. Read `man zshexpn` for the primary example of what I'm talking about then look at the scripts in */usr/share/zsh*. For example, I challenge you to tell me what this line from *_absolute_command_paths* does:

```
local -a matches=( "${(@)commands[(R)${~i}[^/]#]}" )
```

**Comment by krader1961** [comment]
I'll also point out that the first time I read this issue I thought the O.P. was talking about "here documents". That is, a construct such as this:

```
> echo << EOF
hello
EOF
```

The fact it is extremely easy, and likely, for someone to read `<<< word` as a here-doc, rather than a here-string, is another reason not to introduce this syntax.

**Comment by xiaq** [comment]
Yeah, I agree that in its POSIX form, herestrings don't seem to be worth the hassle.

But generalize it to value inputs it can be more useful. To recap, many commands in Elvish take input either via (1) value input, or (2) as a final list argument (see https://elv.sh/ref/builtin.html#supplying-input for examples). If it's very concise to build value inputs, they only need to support the first form; the second form can just be a redirect. Hypothetical example:

```
// Instead of:
each [x]{ + 1 $x } [10 100]
// Write:
each [x]{ + 1 $x } << [10 100]
```

This might just seem to be more verbose, but the advantage is that it doesn't require the implementation of `each` to support 1 or 2 arguments. Another benefit, in terms of readability, is that redirections can appear anywhere, so you can write:

```
each << [10 100] [x]{
  + 1 $x
}
```

This will allow `each` to eventually fully replace `for`. Although TBH, in this example it is still less readable than `for`, so more work on the syntax of input redirection may still be needed.

**Comment by xiaq** [comment]
#991 is also relevant here.

**Comment by krader1961** [comment]
I like the idea of changing the syntax so that functions, like each, which use the `$input-list?` pattern would instead always read from a value pipe -- possibly implicitly created by the `<<` operator. 

What about a single value not enclosed in brackets? That is, would `count << 1` and `count << [1]` be equivalent? The current behavior special-cases strings as a list of characters. So `count abc` and `str:split '' abc | count` are equivalent. But note that `count (float64 1)` is an error since the `float64` type is not iterable. In fact, AFAICT, no builtin types implement the `Iterable` interface as I write this so only strings are special-cased in the context of an `$input-list?` argument. I appreciate why implicitly splitting a string into its constituent code-points is  useful. But note that these two examples are not equivalent:

```
> put abc | count
▶ 1
> count abc
▶ 3
```

As a greybeard I understand why those two statements produce different results. But it is a subtlety that is even more likely to confuse users if `count << abc` is implemented. Should the "here-value" form be equivalent to the first or second example above?
